### ISSUE TITLE
Expose control over embedded Tomcat instance configuration

### ISSUE THREAD

**Issue by yfinkelstein** [line/armeria#83]
I believe the ability to embed Tomcat is a very useful feature of Armenia which will get a big audience once advertised wide. There is lots of old servlet code that people don't want to touch. So, when moving to Netty/Armenia they get a path for smooth transition.

That said, Armenia needs to find a way to expose control over embedded instance creation to the user. As you know, Tomcat has lots of various global config files with a ton of parameters and options. People use them to customize their deployments. Now that Tomcat is embedded, all this is lost - unless a flexible API is exposed to customize it.

In my case I needed 2 things that forced me to modify TomcatServiceInvocationHandler.

1.     private static final String SERVICE_NAME = "Catalina";//"Tomcat-over-Armeria";
The service name ends up being in the ObjectName of every JMX bean that is registered by Tomcat, and JMX is often used to access some internal Tomcat functionality. We had some old code that was expecting service name to be "Catalina" which is the default in Tomcat. Please make the service name configurable! 

2. Thread pools in Tomcat are not created by default. Our code was using one of these resizable StandardThreadExecutor through the JMX API. I had to do this:
```
        StandardThreadExecutor executor = new StandardThreadExecutor();
        executor.setName("DefaultThreadPool");
        service.addExecutor(executor);
```   

Consistent with the spirit of Armenia's service builders I would expose decorators for the following instances:
1. Connector
2. StandardService
3. StandardEngine

and let the user modify everything they desire on these instances. That would accommodate the majority of typical requests that you would receive otherwise :)

**Comment by trustin** [comment]
- +1 for making the service name customizable.
- Armeria currently uses its own thread pool for serving servlet requests (and other blocking tasks.) What happens when you add an executor to a service? Is a servlet/jsp handled in the executor instead of the blocking task executor Armeria uses?

**Comment by trustin** [comment]
@yfinkelstein on a side note, it's very exciting to hear there is an external customer to this project. Please keep your feed back coming! We're very happy to hear and discuss them. :heart:

**Comment by yfinkelstein** [comment]
Creation of a thread pool itself will not cause change in threading model of Armeria. I have been running my servlets without this thread pool successfully. Something else inside of my servlets was failing to initialize but it was our own custom logic that was assuming too much knowledge of the Tomcat environment. I'm guessing that Catalina Connector needs some thread pool to handle incoming IO, and since you're faking it and calling the Service Adapter directly with the request you can get away with your own thread pool. 

Fyi, the default Tomcat distribution comes with conf/server.xml file, and in there, under Service, there is 
```
  <Service name="Catalina">

    <!--The connectors can use a shared executor, you can define one or more named thread pools-->
    <Executor name="tomcatThreadPool" namePrefix="catalina-exec-"
        maxThreads="150" minSpareThreads="4"/>
   ...
```  

I don't know how does this work down the last detail, but I believe somewhere in the connector (and btw there are many of them - Http, Ajp, Native, etc) there is a logic to dispatch the IO to one of these Executor instances.

Also, here https://tomcat.apache.org/tomcat-8.0-doc/config/executor.html
it says: 
```
The executor is a nested element to the Service element. And in order for it to be picked up by the connectors, the Executor element has to appear *prior* to the Connector element in server.xml
```

**Comment by yfinkelstein** [comment]
More importantly, I'm not suggesting that you always create the executor as a part of tomcat creation sequence. I'm simply suggesting to allow the user to customize every details of the Tomcat configuration including the ability to create executors if so desired for some reason.

**Comment by yfinkelstein** [comment]
Regarding your side note, 
I think the project has a big potential. In my mind, it belongs to some place around gRPC and Vert.x (the HTTP Server part of it). Both of these are specialized, but Armeria can become the low-level framework that does not prescribe a particular Http version, RPC payload format, and even allows for both Servlet and async API processing models. If you can evolve the REST API support Vert.x style and make gRPC work as a part of single Armeria Server instance that would really useful to many. 

The following would be my version of "the best of the web" :)

```
        new ServerBuilder()
                .service (PathMapping.ofPrefix ("/admin/", false), TomcatService.forCurrentClassPath())
                .service (PathMapping.ofPrefix ("/grpc/", false), gRPCServiceBuilder.build(..))
                .service (PathMapping.ofPrefix ("/restAPI/svc1/", true), RestServiceBuilder.build(..))
                ...

.port(8080, SessionProtocol.HTTP).build().start();

```

And you're almost there!

**Comment by trustin** [comment]
@yfinkelstein #96 should solve this issue. And thank you very much for your input. We definitely have those in our road map. Please consider joining us by sending a PR (gRPC, REST, ...) ;-)

**Comment by anuraaga** [comment]
Somewhat tangential, but completely randomly I found this issue and sample code for spring-boot to implement graceful shutdown, we'll probably want something similar too

https://github.com/spring-projects/spring-boot/issues/4657

**Comment by trustin** [comment]
@anuraaga I thought we already have graceful shutdown implemented and `TomcatService` should be covered by it as well because it's just a service?

**Comment by trustin** [comment]
@yfinkelstein Should be fixed, now that #96 has been merged. Please feel free to open another issue if that didn't work for you.
